stl/inc/xlocnum

```diff
@@ -251,6 +251,231 @@ __PURE_APPDOMAIN_GLOBAL locale::id numpunct<_Elem>::id;
+                    ++_Groups[_Group_index];
+                }
+            } else if (_Groups[_Group_index] == '\0' || *_First != _Kseparator) {
+                break; // not a group separator, done
+            } else { // add a new group to _Groups string
+                _Groups.push_back('\0');
+                ++_Group_index;
+            }
+        }
+
+        if (_Group_index != 0) {
+            if (_Groups[_Group_index] > '\0') {
+                ++_Group_index; // add trailing group to group count
+            } else {
+                _Bad = true; // trailing separator, fail
+            }
+        }
+
+        const char* _Grouping_iter      = _Grouping.data();
+        const char* const _Grouping_end = _Grouping.data() + _Grouping.size();
+        char _Current_grouping_count    = '\0';
+        while (!_Bad && _Group_index > 0) {
+            if (_Grouping_iter != _Grouping_end) { // keep the last value when _Grouping is exhausted
+                _Current_grouping_count = *_Grouping_iter; // assign the variable at least once
+                ++_Grouping_iter;
+            }
+
+            --_Group_index;
+            if ((_Current_grouping_count > '\0' && _Current_grouping_count != CHAR_MAX)
+                && ((_Group_index > 0 && _Groups[_Group_index] != _Current_grouping_count)
+                    || (_Group_index == 0 && _Groups[_Group_index] > _Current_grouping_count))) {
+                _Bad = true; // bad group size, fail
+            }
+            // group size okay, advance to next test
+        }
+    }
+
+    if (_Parse_hex && _Seendigit && _Significant == 0) {
+        *_Ptr++ = '0'; // save at least one leading digit for hex
+    }
+
+    const char _Decimal_point = _CSTD localeconv()->decimal_point[0];
+    if (_First != _Last && *_First == _Punct_fac.decimal_point()) { // add .
+        *_Ptr++ = _Decimal_point;
+        ++_First;
+    }
+
+    if ((_Parse_hex || *_Phexexp != _ENABLE_V2_BEHAVIOR) && _Significant == 0) { // 0000. so far
+        for (; _First != _Last && *_First == _Zero_wc; _Seendigit = true, (void) ++_First) {
+            --_Power_of_rep_base; // just count leading fraction zeros
+        }
+
+        if (_Power_of_rep_base < 0) { // put one back
+            *_Ptr++ = '0';
+            ++_Power_of_rep_base;
+        }
+    }
+
+    for (size_t _Idx; _First != _Last && (_Idx = _STD _Find_elem(_Atoms, *_First)) < _Offset_digit_end;
+         _Seendigit = true, (void) ++_First) {
+        if (_Significant < _Max_sig_dig) { // save a significant fraction digit
+            *_Ptr++ = _Src[_Idx];
+            ++_Significant;
+        } else if (_Idx > 0) {
+            _Sticky = true; // just update _Sticky
+        }
+    }
+
+    if (_Sticky) { // increment ls digit in memory of those lost
+        char& _Last_got_digit = _Ptr[-1] == _Decimal_point ? _Ptr[-2] : _Ptr[-1];
+        if (_Last_got_digit == '0' || _Last_got_digit == (_Parse_hex ? '8' : '5')) {

```

Comment:
This is a weird, weird behavior. If the digit right after the last got digit was `1`, why are we incrementing this? That's like... ceiling, rather than rounding. Additionally, this is true even if we have something like `0.000|0000000001` (where the `|` is the final significant digit), so we'd get out `0.001`, which seems ... very wrong.

Reply:
I do not feel comfortable with this behavior _without_ at least a semi-rigourous understanding of what this does, why this works, etc, with examples of fields which get converted to distinct floating point values under this.

Reply:
> I don't actually think that the 5/8 behavior is wrong, but I'm very concerned about the 0 behavior.

Hmm... the input may be "1.0000000000000001110223024625156540423631668090820312500... [1000 zeroes] ...0001", which is slightly above the midpoint between 1.0 and `std::nexttoward(1.0, 2.0)`.

I think the conditional ceiling is equivalent to correct rounding, because this has the input cut down with a sufficient length.

Reply:
This is not a rigourous understanding of what this does. I fundamentally don't understand how this works, or what it's supposed to do.

Reply:
One way this could be partially answered is through explicit test cases that pass or fail depending on whether each case is handled.

Reply:
The boundaries between each range of decimal numbers that round to a same 32-bit or 64-bit floating-point number, which are midpoints between floating-point numbers for `FE_TONEAREST`, and floating-point numbers for `FE_TOWARDZERO`, `FE_DOWNWARD` and `FE_UPWARD`, are either integers with <=309 digits or fractionals ending with 5 with <=768 significant digits. Thus their `_MAX_SIG_DIG_V2`th digits are either 0 or 5.

That means decimal numbers strictly between `xxx0` and `xxx5` (`xxx0|aaa`, `xxx1`, `xxx1|bbb`, ... `xxx4` and `xxx4|eee`) always convert to the same floating-point number. The same goes for `xxx5|fff`, `xxx6`, `xxx6|ggg`, ... `xxx9` and `xxx9|jjj`.